File: _eunjin/2257_화학식량.py
string = input()

# 스택 문제
stack = []

for char in string:
    if char == "H":
        stack.append(1)
    elif char == "C":
        stack.append(12)
    elif char == "O":
        stack.append(16)
    elif char == "(":
        stack.append(0)
    elif char == ")":  # 괄호 구간 닫기, 가장 최근 괄호 열린 구간까지의 화학식량 합산
        temp = 0
        while stack and stack[-1] != 0:  # 0이 나올 때까지 pop
            temp += stack.pop()
        stack.pop()  # 0을 pop
        stack.append(temp)  # 괄호 구간의 화학식량 다시 넣기
    else:  # 숫자 나오면 stack의 가장 끝에 담긴 값에 곱하기
        stack.append(stack.pop() * int(char))

print(sum(stack))

File: _chongmyoung/2257_화학식량.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s = br.readLine().trim();

        Deque<Integer> st = new ArrayDeque<>();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            if (ch == '(') {
                st.push(-1); // 구분자
            } else if (ch == ')') {
                int sum = 0;
                while (!st.isEmpty() && st.peek() != -1) sum += st.pop();
                st.pop(); // '(' 제거
                st.push(sum); // 괄호 그룹 합
            } else if (ch == 'H' || ch == 'C' || ch == 'O') {
                int val = (ch == 'H') ? 1 : (ch == 'C') ? 12 : 16;
                st.push(val);
            } else { // 숫자 2~9
                int mul = ch - '0';
                int top = st.pop();
                st.push(top * mul);
            }
        }

        int ans = 0;
        while (!st.isEmpty()) ans += st.pop();
        System.out.println(ans);
    }
}
